File: radixSort.js
function countingSort(arr, div) {
    let n = arr.length
    let out = Array(n).fill(0)
    let count = Array(10).fill(0)

    for (let x of arr) {
        count[Math.floor(x / div) % 10]++
    }

    for (let i = 1; i < 10; i++) {
        count[i] += count[i - 1]
    }
    
    for (let i = n-1; i >= 0; i--) {
        let indx = Math.floor(arr[i] / div) % 10
        out[count[indx] - 1] = arr[i]
        count[indx]--
    }

    for (let i = 0; i < n; i++) {
        arr[i] = out[i]
    }
}

function radixSort(arr) {
    let m = Math.max(...arr)
    let div = 1
    while (Math.round(m / div) > 0) {
        countingSort(arr, div)
        div *= 10
    }
}

if (require.main.children.length === 0) {
    let arr = process.argv.slice(2)
    arr = arr.map(num => parseInt(num))
    arr = arr.filter(num => Number.isInteger(num))
    radixSort(arr)
    console.log(arr)
}

File: radixSort.py
import sys

def countingSort(arr, div):
    n = len(arr)
    out = [0]*n
    count = [0]*10

    for x in arr:
        count[(x // div)%10] += 1
    
    for i in range(1, 10):
        count[i] += count[i - 1]

    for x in arr[::-1]:
        indx = (x // div)%10
        out[count[indx] - 1] = x
        count[indx] -= 1
    
    for i in range(n):
        arr[i] = out[i]

def radixSort(arr):
    m = max(arr)
    div = 1
    while m // div:
        countingSort(arr, div)
        div *= 10

if __name__ == '__main__':
    arr = list(map(int, sys.argv[1:]))
    radixSort(arr)
    print(arr)
